rhombus-lib/rhombus/private/amalgam/filesystem.rhm

```diff
@@ -47,5 +71,344 @@ namespace filesystem:
   fun link_exists(p :: PathString):
     rkt.#{link-exists?}(p)
 
-  fun list_directory(p :: PathString = Path.current_directory()) :~ List.of(Path):
-    List(& rkt.#{directory-list}(p))
+  fun type(p :: PathString,
+           ~must_exist = #false):
+    ~who: who
+    match (rkt_error.local_error_adjust { #'#{file-or-directory-type}: who }:
+             rkt.#{file-or-directory-type}(p))
+    | #'#{directory-link}: #'directory_link
+    | s: s
+
+  fun files(start_p :: PathString = Path.current_directory(),
+            ~recur = #false,
+            ~follow_links = #false,
+            ~skip: skip :: Path -> Any = fun (x): #false,
+            ~add_path = #false) :~ List.of(Path):
+    ~who: who
+    fun files_here(accum :~ List, p, onto_path):
+      let files :~ PairList:
+        rkt_error.local_error_adjust { #'#{directory-list}: who }:
+          rkt.#{directory-list}(p, #{#:build?}: !onto_path && add_path)
+      for values (accum :~ List = accum) (file: files):
+        let file = (if onto_path | Path.add(onto_path, file) | file)
+        skip_when skip(file)
+        let new_accum = accum.add(file)
+        if recur && (if follow_links
+                     | directory_exists(file)
+                     | type(file) == #'directory)
+        | files_here(new_accum, if add_path | file | Path.add(start_p, file), file)
+        | new_accum
+    files_here([], start_p, #false)
+
+  fun roots() :~ List.of(Path.Absolute):
+    ~who: who
+    rkt_error.local_error_adjust { #'#{filesystem-root-list}: who }:
+      List(& rkt.#{filesystem-root-list}())
+
+  fun rename(p :: PathString,
+             to_p :: PathString,
+             ~exists_ok = #false):
+    ~who: who
+    rkt_error.local_error_adjust { #'#{rename-file-or-directory}: who }:
+      rkt.#{rename-file-or-directory}(p, to_p, exists_ok)
+
+  fun delete(p :: PathString,
+             ~as: mode :: matching(#'any || #'file || #'directory) = #'any,
+             ~recur = #false,
+             ~must_exist = #true):
+    ~who: who
+    fun delete(p):
+      fun delete_file():
+        cond
+        | (rkt.#{system-type}() == #'windows) && rkt.#{file-exists?}(p):
+            let tmp = make_temporary()
+            let moved:
+              try:
+                rkt.#{rename-file-or-directory}(p, tmp, #true)
+                #true
+                ~catch (x :: Exn.Fail.Filesystem):
+                  #false
+            rkt_error.local_error_adjust { #'#{delete-file}: who }:
+              unless moved
+              | rkt.#{delete-file}(p)
+              rkt.#{delete-file}(tmp)
+        | ~else:
+            rkt_error.local_error_adjust { #'#{delete-file}: who }:
+              rkt.#{delete-file}(p)
+      cond
+      | (recur && mode != #'file) || mode == #'any:
+          match (rkt_error.local_error_adjust { #'#{file-or-directory-type}: who }:
+                   rkt.#{file-or-directory-type}(p))
```

Comment:
Can this be simplified to use `filesystem.type`?

Reply:
The idea was to have any error reported as `delete`, but maybe this is a case where no error is possible.

Reply:
Oh, that makes more sense.  This `local_error_adjust` pattern with `file-or-directory-type` happens a few times in the module, maybe it could be a private function that takes a `who` parameter?